vllm/multimodal/processing.py

```diff
@@ -212,62 +555,136 @@ def __call__(
+            for modality, matches in full_groupby_modality(token_matches)
+        ):  # yapf: disable
+            token_ids = replace_token_matches(
+                token_ids,
+                token_matches,
+                mm_items,
+                hf_inputs,
+            )
+
+            text = _decode(tokenizer, token_ids)
+            matched_repls = [match.prompt_repl for match in token_matches]
+        else:
+            text = _decode(tokenizer, token_ids)
 
-            modality_placeholders: List[PlaceholderRange] = []
+            text_matches = find_text_matches(text, prompt_repls)
+            text = replace_text_matches(
+                text,
+                text_matches,
+                mm_items,
+                hf_inputs,
+            )
 
-            for item_idx, orig_item in enumerate(orig_inputs):
-                for match_text, replace_fn in placeholder_replacements.items():
-                    candidates = candidate_placeholders(tokenizer, match_text)
-                    get_replacement_ids = partial(
-                        replace_fn,
-                        orig_item,
-                        processed_inputs,
-                        item_idx,
-                    )
+            token_ids = _encode(tokenizer, text)
+            matched_repls = [match.prompt_repl for match in text_matches]
+
+        placeholder_ranges = self._find_placeholders(matched_repls, token_ids)
+
+        return token_ids, text, placeholder_ranges
+
+    def apply(
+        self,
+        prompt_text: str,
+        mm_data: MultiModalDataDict,
+        mm_processor_kwargs: Mapping[str, object],
+    ) -> MultiModalInputsV2:
+        tokenizer = self.ctx.tokenizer
 
-                    for match_ids in candidates:
-                        # TODO(youkaichao): Don't update new_token_ids
-                        placeholders = apply_placeholders(
-                            new_token_ids,
-                            match_ids,
-                            get_replacement_ids,
-                        )
+        hf_inputs = self._apply_hf_processor(prompt_text, mm_data,
+                                             mm_processor_kwargs)
+        prompt_ids, = hf_inputs.pop("input_ids").tolist()
+        mm_kwargs = MultiModalKwargs(hf_inputs)
 
-                        if placeholders is not None:
-                            modality_placeholders.append(placeholders)
+        all_prompt_repls = self._bind_prompt_replacements(mm_data)
 
-            # yapf: disable
-            mm_placeholders[modality] = modality_placeholders  # type: ignore[index]
-            # yapf: enable
+        # In case HF processor already inserts placeholder tokens
+        all_placeholders = self._find_placeholders(all_prompt_repls,
+                                                   prompt_ids)
+
+        # Otherwise, we insert them ourselves
+        if all_placeholders:
+            prompt_text = _decode(tokenizer, prompt_ids)
+        else:
```

Comment:
```suggestion
        if all_placeholders:
            prompt_text = _decode(tokenizer, prompt_ids)

        # Otherwise, we insert them ourselves
        else:
```